dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -0,0 +1,223 @@
+
+            LOG_TRACE(
+                logger,
+                "trying to schedule task of resource group {}, priority: {}, ru exhausted: {}, is_finished: {}, "
+                "task_queue.empty(): {}",
+                group_info.name,
+                group_info.priority,
+                ru_exhausted,
+                is_finished,
+                group_info.task_queue->empty());
+
+            // When highest priority of resource group is less than zero, means RU of all resource groups are exhausted.
+            // Should not take any task from nested task queue for this situation.
+            if (ru_exhausted)
+                break;
+
+            if (group_info.task_queue->empty())
+            {
+                // Nested task queue is empty, continue and try next resource group.
+                size_t erase_num = resource_group_task_queues.erase(group_info.name);
+                RUNTIME_CHECK_MSG(
+                    erase_num == 1,
+                    "cannot erase corresponding TaskQueue for task of resource group {}, erase_num: {}",
+                    group_info.name,
+                    erase_num);
+                resource_group_infos.pop();
+            }
+            else
+            {
+                // Take task from nested task queue, and should always take succeed.
+                // Because this task queue should not be finished inside lock_guard.
+                RUNTIME_CHECK(group_info.task_queue->take(task));
+                assert(task != nullptr);
+                break;
+            }
+        }
+
+        if (task != nullptr)
+            return true;
+
+        // Wakeup when:
+        // 1. finish() is called.
+        // 2. refill_token_callback is called by LAC.
+        cv.wait(lock);
+        updateResourceGroupInfosWithoutLock();
+    }
+}
+
+template <typename NestedTaskQueueType>
+void ResourceControlQueue<NestedTaskQueueType>::updateStatistics(const TaskPtr & task, ExecTaskStatus, size_t inc_value)
+{
+    assert(task);
+    const String & name = task->getResourceGroupName();
+
+    std::lock_guard lock(mu);
+    auto ru = toRU(inc_value);
+    LOG_TRACE(logger, "resource group {} will consume {} RU(or {} cpu time in ns)", name, ru, inc_value);
+    LocalAdmissionController::global_instance->consumeResource(name, ru, inc_value);
+    updateResourceGroupInfosWithoutLock();
+}
+
+template <typename NestedTaskQueueType>
+void ResourceControlQueue<NestedTaskQueueType>::updateResourceGroupInfosWithoutLock()
+{
+    std::priority_queue<ResourceGroupInfo> new_resource_group_infos;
+    while (!resource_group_infos.empty())
+    {
+        const ResourceGroupInfo & group_info = resource_group_infos.top();
+        auto new_priority = LocalAdmissionController::global_instance->getPriority(group_info.name);
+        new_resource_group_infos.push({group_info.name, new_priority, group_info.task_queue});
+        resource_group_infos.pop();
+    }
+    resource_group_infos = new_resource_group_infos;
```

Comment:
```suggestion
    resource_group_infos = std::move(new_resource_group_infos);
```

Reply:
Maybe it would be simpler to use `std::vector<ResourceGroupInfo>` and `std::sort`

Reply:
And then resource_group_infos  can be sorted directly, no need to create new container every time.

Reply:
Using a vector can cause the deletion operation to be more cumbersome

Reply:
ok